### PR TITLE
Fixed a bug that caused request to fail with utf-8 query parameters

### PR DESCRIPTION
**Pull request by czardoz**
Also updated to the latest upstream code, to fix the tests.

### CODE REVIEW

**Comment by czardoz** [comment]
https://github.com/nodejs/node/blob/7f2c9ba1ac343ebdd3d4fbe27757847555bf141d/lib/_http_client.js#L23

File: lib/url-parse.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    parsed.query = qs
+    parsed.search = search
+    parsed.path = path
+
+    str = url.format(parsed)
+  }
+
+  // Parse again, because Node does not guarantee consistency of properties
+  return url.parse(str)
+}
```
Comment (shamasis) [inline lib/url-parse.js]: @czardoz do u want to optimise this core module to use constants? should be easy.

File: lib/url-parse.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    parsed.query = qs
+    parsed.search = search
+    parsed.path = path
+
+    str = url.format(parsed)
+  }
+
+  // Parse again, because Node does not guarantee consistency of properties
+  return url.parse(str)
```
Comment (shamasis) [inline lib/url-parse.js]: explain. :-)

File: package.json [hunk 3]
```diff
@@ -62,11 +62,18 @@
...
     "karma-phantomjs-launcher": "^1.0.0",
-    "karma-tap": "^2.0.1",
+    "karma-tap": "^3.0.1",
     "phantomjs-prebuilt": "^2.1.3",
     "rimraf": "^2.2.8",
     "server-destroy": "^1.0.1",
     "tape": "^4.6.0",
     "taper": "^0.4.0"
+  },
+  "greenkeeper": {
```
Comment (shamasis) [inline package.json]: why?

File: request.js [hunk 4]
```diff
@@ -735,6 +736,11 @@ Request.prototype.start = function () {
   var reqOptions = copy(self)
   delete reqOptions.auth
 
+  // Workaround for a bug in Node: https://github.com/nodejs/node/issues/8321
+  if (!self.proxy && !(self.uri.isUnix)) {
```
Comment (shamasis) [inline request.js]: what is this `self.uri.isUnix`?

File: request.js [hunk 5]
```diff
@@ -959,6 +965,10 @@ Request.prototype.onRequestResponse = function (response) {
     })
 
     responseContent.on('data', function (chunk) {
+      if (self.timing && !self.responseStarted) {
```
Comment (shamasis) [inline request.js]: where did this come from? and why not `Date.now()`?

**Comment by shamasis** [comment]
Also, do you intend to add test for the changes we made?

**Comment by shamasis** [comment]
Seems good to me when integration test and constants are added.

File: lib/url-parse.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  , encode
+
+/**
+ * Percent encodes a query string according to RFC 3986
+ *
+ * @param value
+ * @returns {string}
+ */
+encode = function (value) {
+  return encodeURIComponent(value).replace(/[!'()*]/g, function (c) {
```
Comment (shamasis) [inline lib/url-parse.js]: move this regex as common variable on top

File: lib/url-parse.js (hunk 1 above)
Comment (shamasis) [inline lib/url-parse.js]: You doing this too. Right?

File: request.js (hunk 4 above)
Comment (shamasis) [inline request.js]: I have created an internal tracker task to keep this in view.

File: lib/url-parse.js [hunk 7]
```diff
@@ -1,58 +1,111 @@
...
-      return encode(key) + '=' + encode(value)
-    }).filter(function (x) {
-      return x.length > 0
-    }).join('&') : ''
-  }
+      return result.join('&')
+    }
+
+    return encode(key) + '=' + encode(value)
+  }).filter(function (x) {
```
Comment (shamasis) [inline lib/url-parse.js]: seems like this does not retain empty key-value pairs.

- `a=&b`
- `a&b`
- `&a&b`
- `a&&b`
- `a&b&`

**Comment by czardoz** [comment]
LGTS

**Pull request merged**